File: programers_Lv2/[1차] 캐시.py
def solution(cacheSize, cities):
    answer = 0
    if cacheSize == 0:
        return len(cities) * 5
    cache = [None] * cacheSize
    low_cities = map(lambda x: x.lower(), cities)
    for v in low_cities:
        try:
            idx = cache.index(v)
            cache.append(cache.pop(idx))
            answer += 1
        except:
            answer += 5
            cache = cache[1:]
            cache.append(v)
        
    return answer

print(solution(3, ["Jeju", "Pangyo", "Seoul", "NewYork", "LA", "Jeju", "Pangyo", "Seoul", "NewYork", "LA"]))	# 50
print(solution(2, ["Jeju", "Pangyo", "NewYork", "newyork"])) # 16
print(solution(5, ["leo", "leo", "leo"])) # 7
print(solution(3, ["a", "b", "c", "a"])) # 16
print(solution(3, ["a", "b", "c", "a", "b"])) # 17
print(solution(3, ["a", "b", "c", "a", "b", "d", "a"])) # 23

File: programers_Lv2/[1차] 캐시.js
function solution(cacheSize, cities) {
  let answer = 0;
  if (cacheSize === 0) {
    return cities.length * 5;
  }
  const cache = new Array(cacheSize);
  const lowCities = cities.map((v) => v.toLowerCase());
  lowCities.forEach((v, i) => {
    let idx = cache.indexOf(v);
    if (idx > -1) {
      cache.push(...cache.splice(idx, 1));
      answer++;
    } else {
      answer += 5;
      cache.shift();
      cache.push(v);
    }
  });
  return answer;
}

// console.log(
//   solution(3, [
//     "Jeju",
//     "Pangyo",
//     "Seoul",
//     "NewYork",
//     "LA",
//     "Jeju",
//     "Pangyo",
//     "Seoul",
//     "NewYork",
//     "LA",
//   ])
// ); // 50
console.log(
  solution(3, [
    "Jeju",
    "Pangyo",
    "Seoul",
    "Jeju",
    "Pangyo",
    "Seoul",
    "Jeju",
    "Pangyo",
    "Seoul",
  ])
); // 21
